proposals/feature-dependencies.md

```diff
@@ -0,0 +1,176 @@
+
+Before committing each `round` to the `installationOrder`, sort all elements of round according to **Round Stable Sort***.
+
+Repeat for as many rounds as necessary until the worklist is empty.  If there is ever a round where no elements are added to `installationOrder`, the algorithm should terminate and return an error.  This indicates a circular dependency or other error in the dependency graph.
+
+
+#### Definition: Feature Equality
+
+This specification defines two Features as equal if both Features point to the same exact contents and are executed with the same options.
+
+**For Features published to an OCI registry**, two Feature are identical if their manifest digests are equal, and the options executed against the Feature are equal (compared value by value).  Identical manifest digests implies that the tgz contents of the Feature and its entire `devcontainer-feature.json` are identical.  If any of these conditions are not met, the Features are considered not equal.
+
+**For Features fetched by HTTPS URI**, two Features are identical if the contents of the tgz are identical (hash to the same value), and the options executed against the Feature are equal (compared value by value).  If any of these conditions are not met, the Features are considered not equal.
+
+**For local Features**, each Feature is considered unique and not equal to any other local Feature.
+
+#### Definition: Round Stable Sort
+
+To prevent non-deterministic behavior, the algorithm will sort each **round** according to the following rules:
+
+- Compare and sort each Feature lexiographically by their fully qualified resource name (For OCI-published Features, that means the ID without version or digest.).  If the comparison is equal:
+- Compare and sort each Feature from oldest to newest tag (`latest` being the "most new").  If the comparision is equal:
+- Compare and sort each Feature by their options by:
+  - Greatest number of user-defined options (note omitting an option will default that value to the Feature's default value and is not considered a user-defined option). If the comparison is equal:
+  - Sort the provided option keys lexicographically.  If the comparison is equal:
+  - Sort the provided option values lexicographically. If the comparision is equal:
+- Sort Features by their canonical name (For OCI-published Features, the Feature ID resolved to the digest hash).
+
+If there is no difference based on these comparator rules, the Features are considered equal.
+
+### Comments on existing methods of altering Feature installation order
+
+Two existing properties (1) `installsAfter` on the Feature metadata, and  (2) `overrideFeatureInstallationOrder` in the `devcontainer.json` both exist to alter the installation order of user-defined Features.
+
+:: TODO :: More discussion needed!
```

Comment:
Thoughts on using a preprocessed dependency graph that will allow us to run a single ordering algorithm once to cover `dependsOn`, `installsAfter` and `overrideFeatureInstallOrder` of dependencies:
- `installsAfter` and `overrideFeatureInstallOrder` do not include options, so maybe the dependency graph should have nodes for features and two types of directed edges: one for dependencies with options (`dependenciesWithOptions`) and one for dependencies without options (`dependenciesWithoutOptions`). A node's dependencies will be considered 'installed' and the node ready-to-be-installed when all its `dependenciesWithOptions` are installed and for all of its `dependenciesWithoutOptions` at least one with any set of options is installed.
- The `dependsOn` references are added as `dependenciesWithOptions`.
- The `installsAfter` references that are not going to be installed can be omitted from the preprocessed dependency graph. All others are added as `dependenciesWithoutOptions`.
- `overrideFeatureInstallOrder` can be added to the dependency graph with:
  - Add an artificial node for each feature that is in the `overrideFeatureInstallOrder` list with its `dependenciesWithOptions` referring to all nodes of this feature. (This might be optional, but makes the description and maybe the algorithm simpler.)
  - Let each such artificial node depend on the artifical node of the feature preceeding it in the `overrideFeatureInstallOrder` list. (Enforcing the user-provided order.)
  - Let all leave nodes (nodes without any dependencies) in the graph refer to the artificial node of the feature last in the `overrideFeatureInstallOrder` list. (Enforcing that these are installed first.)
- To float user-selected features (those in the devcontainer.json) to the end of the install order, we could assign each node a priority and in each round of the algorithm only handle nodes of the highest (lowest value) available priority. User-selected features would have priority 2, all others priority 1.
  - To make local features float to the end (as in the existing spec), we could assign them priority 3, so they are installed after user-selected features.

Reply:
master